### PR TITLE
[HW-2]Regex

### PR DESCRIPTION
**Pull request by BukhtikYu**

### CODE REVIEW

File: HW02/Yuliya_Bukhtik/yuliya_bukhtik_hw_02_t_04.rb [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+  # key_2 = 'digits'
+  value2 = string.chars
+
+  # by this method we check if elem is a digit and delete it if it's not
+  value2.delete_if do |elem|
+    elem != elem[/\d+/]
+  end
+
+  Hash['letters', value1.length, 'digits', value2.length]
+end
```
Comment (vladhilko) [inline HW02/Yuliya_Bukhtik/yuliya_bukhtik_hw_02_t_04.rb]: Interesting implementation  👍 

P.S. You can try to read about methods `scan` and `count`

File: HW01/Yuliya_Bukhtik/yuliya_bukhtik_hw_01_t_03.rb [hunk 2]
```diff
@@ -0,0 +1,3 @@
+# frozen_string_literal: true
+
+puts(2222 - 222 + 22 / 2 - 2)
```
Comment (vladhilko) [inline HW01/Yuliya_Bukhtik/yuliya_bukhtik_hw_01_t_03.rb]: Please, remove HW1 tasks from this PR :pensive:

File: HW02/Yuliya_Bukhtik/yuliya_bukhtik_hw_02_t_01.rb [hunk 3]
```diff
@@ -0,0 +1,11 @@
+# frozen_string_literal: true
+
+file_path = 'log_file_task_01.txt'
+
+def task_1(file)
+  IO.foreach(file) do |line|
+    puts line if line.include? 'error'
```
Comment (vladhilko) [inline HW02/Yuliya_Bukhtik/yuliya_bukhtik_hw_02_t_01.rb]: We have to return only the first line with `error`. Your solution returns all lines with errors

File: HW02/Yuliya_Bukhtik/yuliya_bukhtik_hw_02_t_02.rb [hunk 4]
```diff
@@ -0,0 +1,33 @@
+# frozen_string_literal: true
+
+file_path = 'log_file_task_02.txt'
+
+def task_2(file)
+  f = File.new file
+  lines = f.readlines
```
Comment (vladhilko) [inline HW02/Yuliya_Bukhtik/yuliya_bukhtik_hw_02_t_02.rb]: You can do it in one line
```suggestion
  lines = File.readlines(file)
```

File: HW02/Yuliya_Bukhtik/yuliya_bukhtik_hw_02_t_02.rb [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+  # by the below method lines with ERROR will be deleted
+  lines.delete_if { |line| line.include? 'error' }
```
Comment (vladhilko) [inline HW02/Yuliya_Bukhtik/yuliya_bukhtik_hw_02_t_02.rb]: I don't see any line format checks, this method just removes lines with `error`

File: HW02/Yuliya_Bukhtik/yuliya_bukhtik_hw_02_t_03.rb [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+require 'date'
+
+logs = 'log_file_task_03.txt'
+
+def task_3(file)
+  f = File.new file
+  lines = f.readlines
+
+  # getting action duration
+  d = formatted_strings lines
```
Comment (vladhilko) [inline HW02/Yuliya_Bukhtik/yuliya_bukhtik_hw_02_t_03.rb]: can you try to use a more clear name?  `d`  is't readable

File: HW02/Yuliya_Bukhtik/yuliya_bukhtik_hw_02_t_03.rb [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+def formatted_strings(array)
+  # getting only strings with "Calling core with action"
+  array.select! { |elem| elem.include? 'Calling core with action' }
+
+  # getting format date from Calling core with action_strings
+  array.map! { |elem| elem[/\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2}\.\d/] }
+
+  # getting DateTime object from date format
+  a = DateTime.parse(array[0])
+  b = DateTime.parse(array[1])
```
Comment (vladhilko) [inline HW02/Yuliya_Bukhtik/yuliya_bukhtik_hw_02_t_03.rb]: the same comment about the names

File: HW02/Yuliya_Bukhtik/yuliya_bukhtik_hw_02_t_01.rb (hunk 3 above)
Comment (BukhtikYu) [reply HW02/Yuliya_Bukhtik/yuliya_bukhtik_hw_02_t_01.rb]: In telegram this topic was discussed, Tatyan asked if we must return only the first line with error or all lines with error. Alexander asked : "В задании такого требования не было, но если реализуете поиск нескольких строк с ошибками - это будет плюс".

File: HW02/Yuliya_Bukhtik/yuliya_bukhtik_hw_02_t_02.rb (hunk 5 above)
Comment (BukhtikYu) [reply HW02/Yuliya_Bukhtik/yuliya_bukhtik_hw_02_t_02.rb]: Below there is def parsing_new_line(line) that checks line format.

File: HW02/Yuliya_Bukhtik/yuliya_bukhtik_hw_02_t_01.rb (hunk 3 above)
Comment (vladhilko) [reply HW02/Yuliya_Bukhtik/yuliya_bukhtik_hw_02_t_01.rb]: Ok, I didn't see this message

File: HW02/Yuliya_Bukhtik/yuliya_bukhtik_hw_02_t_02.rb [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+
+  # by the below method the rest lines will be parsed into the required format
+  lines.map! do |line|
+    parsing_new_line line
+  end
+
+  f.close
```
Comment (vladhilko) [inline HW02/Yuliya_Bukhtik/yuliya_bukhtik_hw_02_t_02.rb]: Please, delete this line

File: HW02/Yuliya_Bukhtik/yuliya_bukhtik_hw_02_t_02.rb (hunk 5 above)
Comment (vladhilko) [reply HW02/Yuliya_Bukhtik/yuliya_bukhtik_hw_02_t_02.rb]: What if line format is wrong in the log file? Will your code skip these lines? For example:
```txt
1. 2018-04-23 20:30:42: SSL peer: 10.6.246.101, peer cert: , #<Puma::MiniSSL::SSL: System: Undefined: 0 - 0>
2. 1 12 2312 23
3. example line
```

* lines that do not match the format `10.6.246.103 - - [23/Apr/2018:20:30:39 +0300] "POST /test/2/messages HTTP/1.1" 200 48 0.0498` should be ignored

Could you add a check for it?

Comment (vladhilko) [reply HW02/Yuliya_Bukhtik/yuliya_bukhtik_hw_02_t_02.rb]: And this line should be included in the resulting output 

```
10.6.246.101 - - [23/Apr/2018:20:31:39 +0300] "POST /test/2/error HTTP/1.1" 200 48 0.0290
```

**Comment by vladhilko** [comment]
Please, remove HW1 files from this PR and click `Resolve` on every comment if this comment isn't actual anymore

**Pull request merged**